crates/schema/src/def/validate/v8.rs

```diff
@@ -1,9 +1,761 @@
+        custom_ordering: true,
+    }
+}
+
+/// Check a column definition.
+fn check_column(
+    id: ColId,
+    column: RawColumnDefV8,
+    resolved_product_type: &ProductType,
+    table_name: &RawIdentifier,
+    typespace: &Typespace,
+    extra_errors: &mut Vec<ValidationError>,
+) {
+    let RawColumnDefV8 { col_name, col_type } = column;
+
+    // for some reason, the original `RawColumnDefv8` sometimes stored *resolved* types.
+    // so, resolve before checking for equality.
+
+    let element = resolved_product_type.elements.get(id.idx());
+
+    let resolved_col_ty = WithTypespace::new(typespace, &col_type).resolve_refs();
+
+    match (element, resolved_col_ty) {
+        (Some(element), Ok(resolved_col_ty)) => {
+            if !element.has_name(&col_name) || element.algebraic_type != resolved_col_ty {
+                extra_errors.push(ValidationError::ColumnDefMalformed {
+                    column: RawColumnName::new(table_name.clone(), col_name),
+                    ty: resolved_col_ty.into(),
+                    pos: id,
+                    product_type: resolved_product_type.clone().into(),
+                });
+            }
+        }
+        _ => extra_errors.push(ValidationError::ColumnDefMalformed {
+            column: RawColumnName::new(table_name.clone(), col_name),
+            ty: col_type.into(),
+            pos: id,
+            product_type: resolved_product_type.clone().into(),
+        }),
+    }
+}
+
+/// Upgrade an index.
+fn upgrade_index(index: RawIndexDefV8, is_generated: bool) -> RawIndexDefV9 {
+    let RawIndexDefV8 {
+        index_name,
+        is_unique: _, // handled by generated_constraints
+        index_type,
+        columns,
+    } = index;
+
+    let algorithm = match index_type {
+        IndexType::BTree => RawIndexAlgorithm::BTree { columns },
+        IndexType::Hash => RawIndexAlgorithm::Hash { columns },
+    };
+    // The updated bindings macros will correctly distinguish between accessor name and index name as specified in the
+    // ABI stability proposal. The old macros don't make this distinction, so we just reuse the name for them.
+    let accessor_name = if is_generated { None } else { Some(index_name.clone()) };
+    RawIndexDefV9 {
+        name: index_name.clone(),
+        // Set the accessor name to be the same as the index name.
+        accessor_name,
+        algorithm,
+    }
+}
+
+/// Upgrade a constraint.
+fn upgrade_constraint_to_unique_constraint(
+    constraint: RawConstraintDefV8,
+    table_name: &RawIdentifier,
+    primary_key: &mut Option<ColId>,
+    extra_errors: &mut Vec<ValidationError>,
+) -> Option<RawUniqueConstraintDefV9> {
+    let RawConstraintDefV8 {
+        constraint_name,
+        constraints,
+        columns,
+    } = constraint;
+
+    if constraints.has_primary_key() {
+        if let Some(col) = columns.as_singleton() {
+            if primary_key.is_none() {
+                *primary_key = Some(col);
```

Comment:
What's the point of this. Seems unused? Or am I misunderstanding? Is this mutating the argument?

Reply:
Yeah, I'll rewrite this with comments.